ui-lib/src/main/java/co/electriccoin/zcash/work/SyncWorker.kt

```diff
@@ -12,55 +12,115 @@ import cash.z.ecc.android.sdk.Synchronizer
+import kotlin.random.Random
+import kotlin.time.Duration
+import kotlin.time.Duration.Companion.days
 import kotlin.time.Duration.Companion.hours
+import kotlin.time.Duration.Companion.minutes
+import kotlin.time.DurationUnit
+import kotlin.time.toDuration
 import kotlin.time.toJavaDuration
 
+// TODO [#1249]: Add documentation and tests on background syncing
+// TODO [#1249]: https://github.com/Electric-Coin-Company/zashi-android/issues/1249
+
 @Keep
 class SyncWorker(context: Context, workerParameters: WorkerParameters) : CoroutineWorker(context, workerParameters) {
     @OptIn(ExperimentalCoroutinesApi::class)
     override suspend fun doWork(): Result {
-        // Enhancements to this implementation would be:
-        // - Quit early if the synchronizer is null after a timeout period
-        // - Return better status information
+        Twig.debug { "BG Sync: starting..." }
 
         WalletCoordinator.getInstance(applicationContext).synchronizer
             .flatMapLatest {
+                Twig.debug { "BG Sync: synchronizer: $it" }
+
                 it?.status?.combine(it.progress) { status, progress ->
-                    StatusAndProgress(status, progress)
+                    StatusAndProgress(status, progress).also {
+                        Twig.debug { "BG Sync: result: $it" }
+                    }
                 } ?: emptyFlow()
             }
             .takeWhile {
-                it.status != Synchronizer.Status.DISCONNECTED && it.progress.isLessThanHundredPercent()
+                it.status != Synchronizer.Status.DISCONNECTED &&
+                    it.status != Synchronizer.Status.SYNCED
             }
             .collect()
 
+        Twig.debug { "BG Sync: terminating..." }
+
         return Result.success()
     }
 
     companion object {
-        /*
-         * There may be better periods; we have not optimized for this yet.
-         */
-        private val DEFAULT_SYNC_PERIOD = 24.hours
+        private val SYNC_PERIOD = 24.hours
+        private val SYNC_DAY_SHIFT = 1.days // Move to tomorrow
+        private val SYNC_START_TIME_HOURS = 3.hours // Start around 3 a.m. at night
+        private val SYNC_START_TIME_MINUTES = 60.minutes // Randomize with minutes until 4 a.m.
 
         fun newWorkRequest(): PeriodicWorkRequest {
+            val targetTimeDiff = calculateTargetTimeDifference()
+
+            Twig.debug { "BG Sync: necessary trigger delay time: $targetTimeDiff" }
+
             val constraints =
                 Constraints.Builder()
                     .setRequiresStorageNotLow(true)
-                    .setRequiredNetworkType(NetworkType.CONNECTED)
+                    .setRequiredNetworkType(NetworkType.UNMETERED)
+                    .setRequiresCharging(true)
                     .build()
 
-            return PeriodicWorkRequestBuilder<SyncWorker>(DEFAULT_SYNC_PERIOD.toJavaDuration())
+            return PeriodicWorkRequestBuilder<SyncWorker>(SYNC_PERIOD.toJavaDuration())
```

Comment:
The documentation for this constructor says:
> The [`PeriodicWorkRequest`](https://developer.android.com/reference/androidx/work/PeriodicWorkRequest) is guaranteed to run exactly one time during this interval (subject to OS battery optimizations, such as doze mode). [..] It may run immediately, at the end of the period, or any time in between so long as the other conditions are satisfied at the time.

AFAICT this means that the interval will start at some time `X` between 3am and 4am (due to `setInitialDelay`), but that `SyncWorker` can be triggered at any time between `X` and `X + SYNC_PERIOD` (i.e. it could sync at 2pm, if `constraints` happen to be met at that time). Is this the intended behaviour? I don't know how likely it is that jobs are always run at the beginning of their interval.

If we actually want to require that the background sync only be triggered to start between 3am and 4am, then the alternative "flex period" constructor would be more suitable: we'd set the interval to start at 4am, with a 1 hour `flexInterval` so that the flex perid is 3-4am (`repeatInterval - flexInterval` to the end of the interval, per the docs).

Reply:
A side-effect of using the flex period would be that the first background sync would be on the second night after first installing the app (as we can only shift the start point of the interval with `setInitialDelay`, and the flex period is the last `flexInterval` of the interval). But maybe there's a way around that by also scheduling a one-off task for the first night.

Reply:
Thanks for pointing this out. Yes, I was aware of the `flexInterval` when implementing this. It is one option to consider for our use case. The "second night" drawback was something that I didn't know how to by-pass, and from my experience and testing, the trigger usually happened at the beginning of the `repeatInterval` as we have limited it with the **unmetered connection** and **is plugged into the power** constraints under which is the system in a perfect position to start any planned worker. 

My second thought is that we might have only some of the users having their devices plugged in and on wifi at night, so the trigger for those will start once these constraints are met, and that could be at any time during the day. Thus, the 3-4 a.m. requirement would not be met, but the main goal of a regular sync will be met.

We can still file a follow-up and revisit the sync planning again. What do you think?

Reply:
As long as we anticipate user feedback of the form "my wallet isn't syncing" because the sync process is getting triggered when their phone is plugged in for a short time during the day, instead of for a long time at night, then I think this is fine to merge as-is and we can revisit this later if needed.

Reply:
Filed #1258.